src/main/java/com/iota/iri/storage/localinmemorygraph/LocalInMemoryGraphProvider.java

```diff
@@ -441,42 +442,151 @@ void printTopOrder(HashMap<Integer, Set<Hash>> topOrder)
+        if (revGraph.containsKey(block)) {
             return revGraph.get(block);
         }
         return new HashSet<>();
     }
 
     //TODO for debug
-    public static void printScore() {
-        for(Hash key : score.keySet()) {
-            if(nameMap != null) {
-                System.out.print(nameMap.get(key)+":"+score.get(key));
-            } else {
-                System.out.print(key+":"+score.get(key));
-            }
-            System.out.println();
-        }
-    }
-
-    public void deleteBatch(Collection<Pair<Indexable, ? extends Class<? extends Persistable>>> models) throws Exception
-    {
+//    public static void printScore() {
+//        for(Hash key : score.keySet()) {
+//            if(nameMap != null) {
+//                System.out.print(nameMap.get(key)+":"+score.get(key));
+//            } else {
+//                System.out.print(key+":"+score.get(key));
+//            }
+//            System.out.println();
+//        }
+//    }
+
+    public void deleteBatch(Collection<Pair<Indexable, ? extends Class<? extends Persistable>>> models) throws Exception {
         // TODO implement this
     }
 
-    public void clear(Class<?> column) throws Exception
-    {
+    public void clear(Class<?> column) throws Exception {
         // TODO implement this
     }
 
-    public void clearMetadata(Class<?> column) throws Exception
-    {
+    public void clearMetadata(Class<?> column) throws Exception {
         // TODO implement this
     }
 
-    public long getTotalTxns() throws Exception
-    {
+    public long getTotalTxns() {
         long ret = 0;
         return ret;
     }
-} 
+
+    public List<Hash> totalTopOrder() {
+        return confluxOrder(getPivot(getGenesis()));
+    }
+
+    public List<Hash> confluxOrder(Hash block) {
+        LinkedList<Hash> list = new LinkedList<>();
+        if (block == null || !graph.keySet().contains(block)) {
+            return list;
+        }
+        do {
+            Hash parent = parentGraph.get(block);
+            List<Hash> subTopOrder = new ArrayList<>();
+            List<Hash> diff = new ArrayList<>(past(block));
+            diff.removeAll(past(parent));
+            while (diff.size() != 0) {
+                Map<Hash, Set<Hash>> subGraph = buildSubGraph(diff);
+                List<Hash> noBeforeInTmpGraph = subGraph.entrySet().stream().filter(e -> CollectionUtils.isEmpty(e.getValue())).map(Map.Entry::getKey).collect(Collectors.toList());
+                noBeforeInTmpGraph.sort(Comparator.comparingInt((Hash o) -> lvlMap.get(o)).thenComparing(o -> o));
+                subTopOrder.addAll(noBeforeInTmpGraph);
+                diff.removeAll(noBeforeInTmpGraph);
+            }
+            list.addAll(0, subTopOrder);
+            block = parentGraph.get(block);
+        } while (parentGraph.get(block) != null && parentGraph.keySet().contains(block));
+        return list;
+    }
+
+    private Map<Hash, Set<Hash>> buildSubGraph(List<Hash> diff) {
```

Comment:
add test for buildSubGraph, and the parameter should not be named 'diff' which is too ad hoc